### PR TITLE
Feat: LCFS - Reformat In-App Notifications tied to Email Subscriptions into AG Grid

### PR DESCRIPTION
**Pull request by prv-proton**
#1135 
- migration, added transaction_id and type columns to NotificationMessage table.
- update and fixed to test cases.
- create app notifications when statuses change in InitiativeAgreement, transfer and compliance reports.
- Below UI for notifications page
image

### CODE REVIEW

File: backend/lcfs/web/api/initiative_agreement/services.py [hunk 1]
```diff
@@ -208,16 +209,27 @@ async def director_approve_initiative_agreement(
             initiative_agreement.transaction_effective_date = datetime.now().date()
 
         await self.repo.refresh_initiative_agreement(initiative_agreement)
-        await self._perform_notificaiton_call(initiative_agreement)
 
-    async def _perform_notificaiton_call(self, ia, re_recommended=False):
+    async def _perform_notificaiton_call(self, ia, returned=False):
```
Comment (dhaselhan) [inline backend/lcfs/web/api/initiative_agreement/services.py:213]: suggestion: Fix spelling of notification

File: backend/lcfs/web/api/notification/repo.py [hunk 2]
```diff
@@ -76,7 +91,86 @@ async def get_notification_messages_by_user(
...
+                .joinedload(UserProfile.user_roles)
+                .joinedload(UserRole.role),
+            )
+            .where(and_(*conditions))
+        )
+
+        # Apply sorting
+        if not pagination.sort_orders:
+            query = query.order_by(NotificationMessage.create_date.desc())
+        # for order in pagination.sort_orders:
```
Comment (dhaselhan) [inline backend/lcfs/web/api/notification/repo.py:156]: Do we need this commented out code?

File: backend/lcfs/web/api/notification/schema.py [hunk 3]
```diff
@@ -1,11 +1,39 @@
...
 from lcfs.db.models.initiative_agreement.InitiativeAgreementStatus import (
     InitiativeAgreementStatusEnum,
 )
 from lcfs.db.models.transfer.TransferStatus import TransferStatusEnum
-from lcfs.web.api.base import BaseSchema, NotificationTypeEnum
+from lcfs.web.api.base import BaseSchema, NotificationTypeEnum, PaginationResponseSchema
+from pydantic import computed_field, model_validator
+
+
+class OrganizationSchema(BaseSchema):
```
Comment (dhaselhan) [inline backend/lcfs/web/api/notification/schema.py:13]: We should be re-using the one from Organization

File: backend/lcfs/web/api/notification/schema.py [hunk 4]
```diff
@@ -1,11 +1,39 @@
...
+    organization_id: int
+    name: str
+
+
+class UserProfileSchema(BaseSchema):
```
Comment (dhaselhan) [inline backend/lcfs/web/api/notification/schema.py:18]: We should re-use the one from User

File: backend/lcfs/web/api/transfer/services.py [hunk 5]
```diff
@@ -264,37 +264,67 @@ async def update_transfer(self, transfer_data: TransferCreateSchema) -> Transfer
...
+            status=(
+                new_status.status
+                if status_has_changed or re_recommended
+                else "Return to analyst"
+            ),
+        )
         return transfer_result
 
-    async def _perform_notificaiton_call(self, transfer):
+    async def _perform_notificaiton_call(
```
Comment (dhaselhan) [inline backend/lcfs/web/api/transfer/services.py:277]: suggestion: Fix spelling of notification

**Review by dhaselhan — CHANGES_REQUESTED** [review]

File: backend/lcfs/db/models/notification/NotificationMessage.py [hunk 6]
```diff
@@ -32,12 +33,15 @@ class NotificationMessage(BaseModel, Auditable):
     notification_type_id = Column(
         Integer, ForeignKey("notification_type.notification_type_id")
     )
+    transaction_id = Column(Integer, ForeignKey("transaction.transaction_id"), nullable=True)
 
     # Models not created yet
     # related_transaction_id = Column(Integer,ForeignKey(''))
```
[the remark this answers is not in the document]
Comment (prv-proton) [reply backend/lcfs/db/models/notification/NotificationMessage.py:39]: this is handled with transaction_id, no longer required.

File: backend/lcfs/web/api/compliance_report/update_service.py [hunk 7]
```diff
@@ -91,22 +77,43 @@ async def update_compliance_report(
...
-            notifications = COMPLIANCE_REPORT_STATUS_NOTIFICATION_MAPPER.get(
-                new_status.status
-            )
             # Add history record
             await self.repo.add_compliance_report_history(report, self.request.user)
+
+        await self._perform_notificaiton_call(report, current_status)
+        return updated_report
+
+    async def _perform_notificaiton_call(self, cr, status):
```
[the remark this answers is not in the document]
Comment (prv-proton) [reply backend/lcfs/web/api/compliance_report/update_service.py:86]: changed it to report.

File: backend/lcfs/web/api/initiative_agreement/services.py (hunk 1 above)
Comment (prv-proton) [reply backend/lcfs/web/api/initiative_agreement/services.py:213]: fixed

File: backend/lcfs/web/api/notification/schema.py (hunk 4 above)
Comment (dhaselhan) [reply backend/lcfs/web/api/notification/schema.py:18]: Creating several schemas that have slightly varying fields with the same names is not maintainable. Lets rename this to NotificationUserSchema or something similar to indicate its purpose.

File: backend/lcfs/web/api/notification/schema.py (hunk 3 above)
Comment (prv-proton) [reply backend/lcfs/web/api/notification/schema.py:13]: For same reason as below, don't want to expose too much information for this endpoint.

File: backend/lcfs/web/api/transfer/services.py (hunk 5 above)
Comment (prv-proton) [reply backend/lcfs/web/api/transfer/services.py:277]: fixed. :)

File: backend/lcfs/web/api/notification/schema.py (hunk 4 above)
Comment (prv-proton) [reply backend/lcfs/web/api/notification/schema.py:18]: yes, agreed. changed as per suggestion.

File: backend/lcfs/web/api/notification/schema.py (hunk 3 above)
Comment (prv-proton) [reply backend/lcfs/web/api/notification/schema.py:13]: changed as per suggestion.

**Review by dhaselhan — APPROVED** [review]
LGTM 👍

File: backend/lcfs/web/api/notification/repo.py (hunk 2 above)
Comment (prv-proton) [reply backend/lcfs/web/api/notification/repo.py:156]: thanks, missed to add sorting logic here. updated code.

**Pull request merged**